apps/astarte_appengine_api/lib/astarte_appengine_api/groups/queries.ex

```diff
@@ -17,115 +17,58 @@
 #
 
 defmodule Astarte.AppEngine.API.Groups.Queries do
+  alias Astarte.AppEngine.API.Groups.GroupedDevice, as: GroupedDevice
   alias Astarte.AppEngine.API.Groups.Group
   alias Astarte.AppEngine.API.Device.DeviceStatus
   alias Astarte.AppEngine.API.Device.DevicesList
+  alias Astarte.AppEngine.API.Device.DeletionInProgress
+  alias Astarte.AppEngine.API.Devices.Device, as: DataBaseDevice
+  alias Astarte.AppEngine.API.Realm, as: DataAccessRealm
   alias Astarte.Core.CQLUtils
   alias Astarte.AppEngine.API.Config
   alias Astarte.Core.Device
   alias Astarte.Core.Realm
+  alias Astarte.AppEngine.API.Repo
+  alias Ecto.Changeset
 
   require Logger
+  import Ecto.Query
 
-  def create_group(realm_name, group_changeset) do
-    Xandra.Cluster.run(:xandra, fn conn ->
-      with {:ok, %Group{devices: devices, group_name: group_name} = group} <-
-             Ecto.Changeset.apply_action(group_changeset, :insert),
-           :ok <- check_all_devices_exist(conn, realm_name, devices),
-           {:group_exists?, false} <-
-             {:group_exists?, group_exists?(conn, realm_name, group_name)},
-           :ok <- add_to_group(conn, realm_name, group_name, devices) do
-        {:ok, group}
-      else
-        {:group_exists?, true} ->
-          {:error, :group_already_exists}
-
-        {:error, {:device_not_found, device_id}} ->
-          error_changeset =
-            group_changeset
-            |> Ecto.Changeset.add_error(:devices, "must exist (#{device_id} not found)")
-
-          {:error, error_changeset}
+  def list_devices(realm_name, group_name, opts \\ []) do
+    keyspace = DataAccessRealm.keyspace_name(realm_name)
 
-        {:error, %Ecto.Changeset{} = error_changeset} ->
-          {:error, error_changeset}
-      end
-    end)
+    if(opts[:details],
+      do: list_devices_with_details(keyspace, group_name, opts),
+      else: list_grouped_devices(keyspace, group_name, opts)
+    )
   end
 
-  def list_groups(realm_name) do
-    Xandra.Cluster.run(:xandra, fn conn ->
-      query = "SELECT DISTINCT group_name FROM :keyspace.grouped_devices"
-
-      with {:ok, prepared} <- prepare_with_realm(conn, realm_name, query),
-           {:ok, %Xandra.Page{} = page} <- Xandra.execute(conn, prepared) do
-        {:ok, Enum.map(page, fn %{"group_name" => group_name} -> group_name end)}
-      else
-        {:error, reason} ->
-          _ = Logger.error("Database error: #{inspect(reason)}.", tag: "db_error")
-          {:error, :database_error}
-      end
-    end)
-  end
+  defp list_devices_with_details(keyspace, group_name, opts) do
+    limit = opts[:limit]
+    query = list_devices_with_details_query(keyspace, group_name, opts) |> limit(^limit)
```

Comment:
unfortunately this does not suffice: manually setting the limit to `nil` is not the same as it not being set:

```elixir
iex(4)> from(Name).limit
nil
iex(5)> (from(Name) |> limit(nil)).limit
%Ecto.Query.LimitExpr{
  expr: nil,
  file: "iex",
  line: 5,
  with_ties: false,
  params: []
}
```

you can see this in the generated sql:

```elixir
iex(7)> Repo.to_sql(:all, from(Name))
{"SELECT object_name, object_type, object_uuid FROM names", []}
iex(8)> Repo.to_sql(:all, from(Name) |> limit(nil))
{"SELECT object_name, object_type, object_uuid FROM names LIMIT NULL", []}
```

this did create problems, at least in the device's tests in ci.

```suggestion
      query = list_devices_with_details_query(keyspace, group_name, opts)
      query = 
        case Keyword.fetch(opts, :limit) do
          {:ok, limit} -> query |> limit(limit)
          :error -> query
        end
```